### PR TITLE
chore: Abuse deftemplate to get rid of Address

### PR DESCRIPTION
**Pull request by jacereda**
I don't remember if this has been discussed before. If you think it's worth it I can polish the patch (it emits a double semicolon right now and I don't know where to place the definition).

### CODE REVIEW

**Comment by hellerve** [comment]
That’s pretty smart!

**Comment by eriksvedang** [comment]
I actually with something similar in #1097, but for dereferencing pointers (adding `*` in front, instead). I think it's worth pursuing, making the compiler smaller and all.

**Comment by eriksvedang** [comment]
I think we need some good tests for it though, since the address operator is not actually used much in the tests code, iirc.

**Comment by jacereda** [comment]
Where should the `deftemplate` go?

**Comment by jacereda** [comment]
Any idea on how to test it?

**Comment by TimDeve** [comment]
Being a C macro it will break if you try to pass address as a ref callback, no? Doesn't address have to be a built-in instead of a function?

**Comment by hellerve** [comment]
Sidenote: currently `address` will also break then, because it is a special form.

**Comment by TimDeve** [comment]
Do you not get a compiler error if you try to use a built-in/special form as a function?

**Comment by hellerve** [comment]
Yes, but a rather unfortunate one:

```
> (Array.copy-map &address &[1 2 3])
I didn’t understand the form `address` at line 1, column 18 in 'REPL'.

Is it valid? at REPL:1:1.
```

I guess this is better than the error you get when using the `deftemplate` version, which generates invalid C. Sadly just making a template function might not work either, due to value semantics.

**Comment by scolsen** [comment]
> Sadly just making a template function might not work either, due to value semantics.

What do you mean by value semantics in this context--are you referring to how the borrow checker might handle this? I think a template function should work ok based on some of the templates in `Pointer.carp`, but I'm probably missing a nuance to borrowing/ref management--I suppose the Ptr's in the Array would be unowned, but I think that's ok?

**Comment by jacereda** [comment]
We could change `address` to receive a `Ref`, that way it could be a plain function instead of a macro.

**Comment by scolsen** [comment]
Looks like there needs to be some changes to the SDL sources too, but otherwise this should work and I like this approach! `address` only taking a Ref makes sense to me.

**Comment by hellerve** [comment]
> > Sadly just making a template function might not work either, due to value semantics.
> 
> What do you mean by value semantics in this context--are you referring to how the borrow checker might handle this? I think a template function should work ok based on some of the templates in `Pointer.carp`, but I'm probably missing a nuance to borrowing/ref management--I suppose the Ptr's in the Array would be unowned, but I think that's ok?

The problem is with the value semantics of C: you can’t take the address of a stack variable; `Int`, for instance, would be one of those.

**Comment by scolsen** [comment]
> > > Sadly just making a template function might not work either, due to value semantics.
> > 
> > 
> > What do you mean by value semantics in this context--are you referring to how the borrow checker might handle this? I think a template function should work ok based on some of the templates in `Pointer.carp`, but I'm probably missing a nuance to borrowing/ref management--I suppose the Ptr's in the Array would be unowned, but I think that's ok?
> 
> The problem is with the value semantics of C: you can’t take the address of a stack variable; `Int`, for instance, would be one of those.

Gotcha thanks! I guess that should be taken care of via `blit` and checking that whatever is passed to `address` does not implement `blit`--but right now that would again limit us to handling `address` in the compiler unless we add some general mechanism for expressing "assert this interface isn't implemented" as discussed in #1095

**Comment by jacereda** [comment]
I assumed the use case for `address` was always related to interfacing with C and the current state of the PR (plain function taking a `Ref`) seems to handle that. Do you have examples where this wouldn't work?

**Comment by scolsen** [comment]
> I assumed the use case for `address` was always related to interfacing with C and the current state of the PR (plain function taking a `Ref`) seems to handle that. Do you have examples where this wouldn't work?

That's my interpretation too. I think `Unsafe.leak` is to make something unmanaged.

**Comment by TimDeve** [comment]
Ok, works for me, hadn't seen the update . Are we saying it's alright to make Ptr `blit` because it's unmanaged?

File: core/Blitable.carp [hunk 1]
```diff
@@ -57,3 +57,7 @@
 (defmodule Uint64
   (defn blit [x] (the Uint64 x))
   (implements blit Uint64.blit))
+
+(defmodule Pointer
```
Comment (eriksvedang) [inline core/Blitable.carp:61]: Why is this instance needed?

Comment (jacereda) [reply core/Blitable.carp:61]: `test/address.carp` won't build unless we have this in place IIRC.

**Comment by eriksvedang** [comment]
I had some idea for a tricky situation where this approach maybe wouldn't work, but I can't for my life remember what it was. Probably it wasn't a problem after all, so let's put it in. You mentioned you wanting to clean it up @jacereda, though it looks totally fine to me. Should we just merge it as is?

**Comment by jacereda** [comment]
IIRC the cleanup part was referring to removing the double semicolon and I think https://github.com/carp-lang/Carp/pull/1111/commits/6e917d5dae5f8c44449a7dbe02353d4638a3661d handled that.

**Pull request closed without merging**